python/cuml/svm/svm.pyx

```diff
@@ -0,0 +1,297 @@
+        # The CppSVC class manages the memory of the parameters that are found
+        # during fitting (the support vectors, and the dual_coefficients). The
+        # number of these parameters are not known before fitting.
+        int n_support
+        math_t *dual_coefs
+        int *support_idx
+        math_t b
+        KernelParams kernel_params
+        math_t C
+        math_t tol
+
+        CppSVC(cumlHandle& handle, math_t C, math_t tol,
+               KernelParams kernel_params, float cache_size,
+               int max_iter) except+
+        void fit(math_t *input, int n_rows, int n_cols, math_t *labels) except+
+        void predict(math_t *input, int n_rows, int n_cols,
+                     math_t *preds) except+
+
+
+class SVC(Base):
+    """
+    SVC (C-Support Vector Classification)
+
+    Currently only binary classification is supported.
+
+    Parameters
+    ----------
+    handle : cuml.Handle
+        If it is None, a new one is created for this class
+    C : float (default = 1.0)
+        Penalty parameter C
+    kernel : string (default='linear' TODO change to 'rbf' once implemented)
+        Specifies the kernel function. Possible options: 'linear', 'poly',
+        'rbf', 'sigmoid', 'precomputed'
+    degree : int (default=3)
+        Degree of polynomial kernel function.
+    gamma : float (default = 'auto')
+        Coefficient for rbf, poly, and sigmoid kernels.
+    coef0 : float (default = 0.0)
+        Independent term in kernel function, only signifficant for poly and
+        sigmoid
+    tol : float (default = 1e-3)
+        Tolerance for stopping criterion.
+    cache_size : float (default = 200 MiB)
+        Size of the kernel cache n MiB (TODO)
+    max_iter : int (default = 100*n_samples)
+        Limit the number of outer iterations in the solver
+    verbose : bool (default = False)
+        verbose mode
+
+    Attributes
+    ----------
+
+    n_support_ : int
+        The total number of support vectors.
+        TODO change this to represent number support vectors for each class.
+    intercept_ : int
+        The constant in the decision function
+
+    The solver uses the SMO method similarily to ThunderSVM and OHD-SVM.
+    """
+    def __init__(self, handle=None, C=1, kernel='linear', degree=3,
+                 gamma='auto', coef0=0.0, tol=1e-3, cache_size=200.0,
+                 max_iter=-1, verbose=False):
+        super(SVC, self).__init__(handle=handle, verbose=verbose)
+        self.tol = tol
+        self.C = C
+        self.kernel = self._get_c_kernel(kernel)
+        self.degree = degree
+        self.gamma = gamma
+        self.coef0 = coef0
+        self.cache_size = cache_size
+        self.max_iter = max_iter
+        self.dual_coefs_ = None  # TODO populate this after fitting
+        self.intercept_ = None
+        self.n_support_ = None
+        self.dtype = None
+        self.kernel_params = None
+        self.svcHandle = None
```

Comment:
We have been avoiding this design because it adds unecessary complexity and is non-trivial to maintain. The SVM should provide completely stateless functions and allow the optimized parameters to be output from `fit()` (using pointer args) and input to `predict()`

Since the parameters live on device, the pickling becomes very easy (and examples can be copied from most other cython wrappers).

Reply:
Using stateless functions now. I do not think it decreased the complexity, we still have to manage the lifetime of the parameter arrays (dual_coef_, support_ & support_vectors_). But that would improve once we are able to add a proper finalizer to these arrays. @dantegd was proposing that with rmm one could add such finalizer. Is there a general solution that could finalize arrays with any allocator?